### PR TITLE
itest: continued itest refactor and fix - IV

### PR DESCRIPTION
**Pull request by yyforyongyu**
Please check #6825 for more context.

Depending on #6823, this PR continues the fix of the tests.

This PR starts at commit `lntemp+itest: refactor testSwitchCircuitPersistence`.

### CODE REVIEW

File: lntest/itest/lnd_zero_conf_test.go [hunk 1]
```diff
@@ -286,110 +230,89 @@ func testOptionScidAlias(net *lntest.NetworkHarness, t *harnessTest) {
...
+	defer func() {
+		// TODO(yy): remove the sleep once the following bug is fixed.
+		// When the payment is reported as settled by Bob, it's
+		// expected the commitment dance is finished and all subsequent
+		// states have been updated. Yet we'd receive the error `cannot
+		// co-op close channel with active htlcs` or `link failed to
+		// shutdown` if we close the channel. We need to investigate
+		// the order of settling the payments and updating commitments
+		// to understand and fix.
+		time.Sleep(2 * time.Second)
```
Comment (GeorgeTsagk) [inline lntest/itest/lnd_zero_conf_test.go:295]: Could there be an alternative to `Sleep`, perhaps waiting for some state change with `Eventually` ?

File: lntest/itest/lnd_zero_conf_test.go [hunk 2]
```diff
@@ -1046,25 +837,49 @@ func testOptionScidUpgrade(net *lntest.NetworkHarness, t *harnessTest) {
...
+	ht.CompletePaymentRequests(bob, []string{daveInvoice2.PaymentRequest})
+
+	// TODO(yy): remove the sleep once the following bug is fixed.  When
+	// the payment is reported as settled by Bob, it's expected the
+	// commitment dance is finished and all subsequent states have been
+	// updated. Yet we'd receive the error `cannot co-op close channel with
+	// active htlcs` or `link failed to shutdown` if we close the channel.
+	// We need to investigate the order of settling the payments and
+	// updating commitments to understand and fix.
+	time.Sleep(2 * time.Second)
```
Comment (GeorgeTsagk) [inline lntest/itest/lnd_zero_conf_test.go:863]: same

**Review by GeorgeTsagk — COMMENTED** [review]
Very nice refactoring! Tests are very clean👌

File: .golangci.yml [hunk 3]
```diff
@@ -78,6 +78,23 @@ linters-settings:
     # so no return split required.
     block-size: 3
 
+  gomnd:
+    # List of numbers to exclude from analysis.
+    # The numbers should be written as string.
+    # Values always ignored: "1", "1.0", "0" and "0.0"
+    # Default: []
+    ignored-numbers:
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply .golangci.yml:86]: yeah for file permissions we allow them to use directly in the code. removed `42` tho, and addresses this change in the previous PR.

File: lntemp/harness_miner.go [hunk 4]
```diff
@@ -376,3 +397,42 @@ func (h *HarnessMiner) GetNumTxsFromMempool(n int) []*wire.MsgTx {
...
 }
+
+// NewMinerAddress creates a new address for the miner and asserts.
+func (h *HarnessMiner) NewMinerAddress() btcutil.Address {
+	addr, err := h.NewAddress()
+	require.NoError(h, err, "failed to create new miner address")
+	return addr
+}
+
+// MineBlocksWithTxes mines a single block to include the specifies
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness_miner.go:408]: fixed in the previous PR.

File: funding/manager.go [hunk 5]
```diff
@@ -1482,6 +1482,10 @@ func (f *Manager) handleFundingOpen(peer lnpeer.Peer,
 		return
 	}
 
+	log.Debugf("Initialized channel reservation: zeroConf=%v, psbt=%v, "+
+		"cannedShim=%v ", reservation.IsZeroConf(),
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply funding/manager.go:1486]: fixed!

File: lntemp/harness.go [hunk 6]
```diff
@@ -466,17 +476,47 @@ func (h *HarnessTest) SuspendNode(node *node.HarnessNode) func() error {
...
+		}
+		h.WaitForBlockchainSync(node)
+
+		return nil
 	}
 }
 
 // RestartNode restarts a given node and asserts.
-func (h *HarnessTest) RestartNode(hn *node.HarnessNode,
+func (h *HarnessTest) RestartNode(hn *node.HarnessNode) {
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness.go:489]: changed in the previous pr.

File: lntemp/harness.go [hunk 7]
```diff
@@ -738,32 +901,108 @@ func (h *HarnessTest) OpenChannelAssertPending(srcNode,
...
+// OpenChannelNoAnnounce attempts to open a channel with the specified
+// parameters extended from Alice to Bob. Additionally, for public channels, it
+// will mine extra blocks so they are announced to the network. In specific,
+// the following items are asserted,
+//   - for non-zero conf channel, 1 blocks will be mined to confirm the funding
+//     tx.
 //   - both nodes should see the channel edge update in their network graph.
 //   - both nodes can report the status of the new channel from ListChannels.
+//   - extra blocks are mined if it's a public channel.
 func (h *HarnessTest) OpenChannel(alice, bob *node.HarnessNode,
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness.go:944]: fixed!

File: lntemp/harness.go [hunk 8]
```diff
@@ -778,13 +1017,49 @@ func (h *HarnessTest) OpenChannel(alice, bob *node.HarnessNode,
...
+	h.AssertTopologyChannelOpen(bob, fundingChanPoint)
+
+	// Finally, check that the channel can be seen in their ListChannels.
+	h.AssertChannelExists(alice, fundingChanPoint)
+	h.AssertChannelExists(bob, fundingChanPoint)
 
 	return fundingChanPoint
 }
 
+// OpenChannelAssertErr opens a channel between node srcNode and dedestNode,
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness.go:1045]: fixed in the previous pr.

File: lntemp/harness_assertion.go [hunk 9]
```diff
@@ -670,6 +709,50 @@ func (h *HarnessTest) AssertNumUTXOs(hn *node.HarnessNode,
 	return h.AssertNumUTXOsWithConf(hn, num, math.MaxInt32, 0)
 }
 
+// getUTOXs gets the number of newly created UTOXs within the current test
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness_assertion.go:712]: my bad🙈 fixed!

File: lntemp/harness_assertion.go [hunk 10]
```diff
@@ -670,6 +709,50 @@ func (h *HarnessTest) AssertNumUTXOs(hn *node.HarnessNode,
...
+
+// GetUTXOs returns all the UTXOs for the given node's account, including
+// confirmed and unconfirmed.
+func (h *HarnessTest) GetUTXOs(hn *node.HarnessNode,
+	account string) []*lnrpc.Utxo {
+
+	return h.getUTXOs(hn, account, math.MaxInt32, 0)
+}
+
+// GetUTXOs returns the confirmed UTXOs for the given node's account.
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness_assertion.go:742]: fixed!

File: lntemp/harness_assertion.go [hunk 11]
```diff
@@ -670,6 +709,50 @@ func (h *HarnessTest) AssertNumUTXOs(hn *node.HarnessNode,
...
+func (h *HarnessTest) GetUTXOsConfirmed(hn *node.HarnessNode,
+	account string) []*lnrpc.Utxo {
+
+	return h.getUTXOs(hn, account, math.MaxInt32, 1)
+}
+
+// GetUTXOs returns the unconfirmed UTXOs for the given node's account.
+func (h *HarnessTest) GetUTXOsUnconfirmed(hn *node.HarnessNode,
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness_assertion.go:750]: fixed!

File: lntemp/harness_miner.go [hunk 12]
```diff
@@ -376,3 +397,42 @@ func (h *HarnessMiner) GetNumTxsFromMempool(n int) []*wire.MsgTx {
...
+// transactions only.
+func (h *HarnessMiner) MineBlockWithTxes(txes []*btcutil.Tx) *wire.MsgBlock {
+	var emptyTime time.Time
+
+	// Generate a block.
+	b, err := h.GenerateAndSubmitBlock(txes, -1, emptyTime)
+	require.NoError(h, err, "unable to mine blocks")
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness_miner.go:415]: fixed in the previous pr

File: lntemp/rpc/lnd.go [hunk 13]
```diff
@@ -319,3 +355,285 @@ func (h *HarnessRPC) ChannelAcceptor() (AcceptorClient, context.CancelFunc) {
 
 	return resp, cancel
 }
+
+// SendCoins sends a given amount of money to the specified address from the
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/rpc/lnd.go:359]: hmmm i think bitcoin is money😂

File: lntemp/rpc/lnd.go [hunk 14]
```diff
@@ -319,3 +355,285 @@ func (h *HarnessRPC) ChannelAcceptor() (AcceptorClient, context.CancelFunc) {
...
+	return resp
+}
+
+// GetRecoveryInfo uses the specifies node to make a RPC call to
+// GetRecoveryInfo and asserts.
+func (h *HarnessRPC) GetRecoveryInfo() *lnrpc.GetRecoveryInfoResponse {
+	ctxt, cancel := context.WithTimeout(h.runCtx, DefaultTimeout)
+	defer cancel()
+
+	req := &lnrpc.GetRecoveryInfoRequest{}
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/rpc/lnd.go:432]: good call. changed in the previous pr.

File: lntemp/rpc/lnd.go [hunk 15]
```diff
@@ -319,3 +355,285 @@ func (h *HarnessRPC) ChannelAcceptor() (AcceptorClient, context.CancelFunc) {
...
+
+	return routes
+}
+
+// SendToRoute makes a RPC call to SendToRoute and asserts.
+func (h *HarnessRPC) SendToRoute() lnrpc.Lightning_SendToRouteClient {
+	// SendToRoute needs to have the context alive for the entire test case
+	// as the returned client will be used for send and receive payment
+	// stream. Thus we use runCtx here instead of a timeout context.
+	client, err := h.LN.SendToRoute(h.runCtx)
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/rpc/lnd.go:484]: it returns the `client` to which you pass your params. and yeah, it's deprecated, we may want to remove the rpc altogether some day.

File: lntest/itest/lnd_channel_policy_test.go [hunk 16]
```diff
@@ -292,22 +236,14 @@ func testUpdateChannelPolicy(net *lntest.NetworkHarness, t *harnessTest) {
...
-		t.Fatalf("expected payment to succeed, instead got %v",
-			sendResp.PaymentError)
-	}
+	require.NoError(ht, err, "unable to receive payment stream")
+	require.Empty(ht, sendResp.PaymentError, "expected payment to succeed")
 
-	// With our little cluster set up, we'll update the fees and the max htlc
-	// size for the Bob side of the Alice->Bob channel, and make sure
+	// With our little cluster set up, we'll update the fees and the max
+	// htlc size for the Bob side of the Alice->Bob channel, and make sure
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_channel_policy_test.go:246]: I think we have a pretty relaxed rule re the spelling of it in comments, only when it comes to actual code we would require a particular style(still have to argue `HTLC` vs `Htlc` tho).

File: lntest/itest/lnd_hold_persistence_test.go [hunk 17]
```diff
@@ -110,97 +85,42 @@ func testHoldInvoicePersistence(net *lntest.NetworkHarness, t *harnessTest) {
...
+		req := &routerrpc.SendPaymentRequest{
+			PaymentRequest: payReq,
+			TimeoutSeconds: 60,
+			FeeLimitSat:    1000000,
 		}
 
-		if payment.Status != lnrpc.Payment_IN_FLIGHT {
-			t.Fatalf("state not in flight: %v", payment.Status)
-		}
+		// Wait for inlight status update.
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_hold_persistence_test.go:113]: fixed in the previous pr

File: lntest/itest/lnd_misc_test.go [hunk 18]
```diff
@@ -207,167 +147,108 @@ func testSphinxReplayPersistence(net *lntest.NetworkHarness, t *harnessTest) {
...
-		assertAmountSent(0, carol, dave), 3*time.Second,
-	)
-	if err != nil {
-		t.Fatalf(err.Error())
-	}
+	ht.AssertAmountPaid("carol => dave", carol, chanPoint, 0, 0)
+	ht.AssertAmountPaid("dave <= carol", dave, chanPoint, 0, 0)
+
+	// Before we restart Dave, make sure both Carol and Dave have added the
+	// HTCL.
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_misc_test.go:204]: fixed in the previous pr

File: lntest/itest/lnd_misc_test.go [hunk 19]
```diff
@@ -871,284 +570,162 @@ func testRejectHTLC(net *lntest.NetworkHarness, t *harnessTest) {
...
-	closeChannelAndAssert(t, net, net.Alice, chanPointAlice, false)
-	closeChannelAndAssert(t, net, carol, chanPointCarol, false)
+	ht.CloseChannel(alice, chanPointAlice)
+	ht.CloseChannel(carol, chanPointCarol)
 }
 
-func testNodeSignVerify(net *lntest.NetworkHarness, t *harnessTest) {
-	ctxb := context.Background()
-
+// testNodeSignVerify checks that only connected nodes are allow to perform
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_misc_test.go:636]: fixed in the previous pr

File: lntest/itest/lnd_misc_test.go [hunk 20]
```diff
@@ -871,284 +570,162 @@ func testRejectHTLC(net *lntest.NetworkHarness, t *harnessTest) {
...
+	ht.CloseChannel(alice, aliceBobCh)
 }
 
-// testAbandonChannel abandones a channel and asserts that it is no
-// longer open and not in one of the pending closure states. It also
-// verifies that the abandoned channel is reported as closed with close
-// type 'abandoned'.
-func testAbandonChannel(net *lntest.NetworkHarness, t *harnessTest) {
-	ctxb := context.Background()
+// testAbandonChannel abandones a channel and asserts that it is no longer open
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_misc_test.go:682]: fixed in the previous pr

File: lntest/itest/lnd_zero_conf_test.go (hunk 1 above)
Comment (yyforyongyu) [reply lntest/itest/lnd_zero_conf_test.go:295]: See [previous discussion](https://github.com/lightningnetwork/lnd/pull/6776#discussion_r957880737). In short, we want to expose it here so we don't forget to fix it.

**Review by GeorgeTsagk — APPROVED** [review]
LGTM 👌🎉

2 tests seem to be failing, one is a btcd unit-race and the other is a timeout while waiting for channel policy update, probably flakes

File: .golangci.yml [hunk 21]
```diff
@@ -94,6 +94,14 @@ linters-settings:
       - 'math.*'
       - 'strconv.ParseInt'
 
+  dupword:
```
Comment (positiveblue) [inline .golangci.yml:97]: q: I have noticed that this linter caught more than one error from when we format the comments to do not have more than 80 columns per line. Any reason to constraint it to the three common words that we use here?

**Review by positiveblue — COMMENTED** [review]
@yyforyongyu the new tests look so much better, and  with ~2k less lines of code!

I have a couple of questions but nothing blocking this. I will resolve some of the conversations because you fixed them in following commits :+1:

File: lntest/itest/lnd_switch_test.go [hunk 22]
```diff
@@ -1134,3 +937,165 @@ func testSwitchOfflineDeliveryOutgoingOffline(
...
+		chanAmt = btcutil.Amount(1000000)
+		pushAmt = btcutil.Amount(900000)
+	)
+
+	params := lntemp.OpenChannelParams{
+		Amt:     chanAmt,
+		PushAmt: pushAmt,
+	}
+
+	// Open a channel with 100k satoshis between Alice and Bob with Alice
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_switch_test.go:385]: updated!

File: lntest/itest/lnd_switch_test.go [hunk 23]
```diff
@@ -1134,3 +937,165 @@ func testSwitchOfflineDeliveryOutgoingOffline(
...
+	// being the sole funder of the channel.
+	alice, bob := ht.Alice, ht.Bob
+
+	// As preliminary setup, we'll create two new nodes: Carol and Dave,
+	// such that we now have a 4 ndoe, 3 channel topology. Dave will make
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_switch_test.go:390]: fixed!

File: lntest/itest/lnd_switch_test.go [hunk 24]
```diff
@@ -176,214 +175,50 @@ func testSwitchOfflineDelivery(ht *lntemp.HarnessTest) {
...
+	// and 5 incoming HTLCs from Alice.
+	ht.AssertNumActiveHtlcs(s.dave, numPayments)
 
-		predErr = assertNumActiveHtlcsChanPoint(dave, carolFundPoint, 0)
-		return predErr == nil
-	}, defaultTimeout)
-	if err != nil {
-		t.Fatalf("htlc mismatch: %v", predErr)
-	}
+	// predErr = assertNumActiveHtlcsChanPoint(dave, carolFundPoint, 0)
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_switch_test.go:201]: Good catch! deleted.

File: lntest/itest/lnd_switch_test.go [hunk 25]
```diff
@@ -176,214 +175,50 @@ func testSwitchOfflineDelivery(ht *lntemp.HarnessTest) {
...
 	// Finally, restart dave who received the settles, but was unable to
 	// deliver them to Alice since they were disconnected.
-	if err := net.RestartNode(dave, nil); err != nil {
-		t.Fatalf("unable to restart dave: %v", err)
-	}
-	if err = net.RestartNode(net.Alice, nil); err != nil {
-		t.Fatalf("unable to restart alice: %v", err)
-	}
+	ht.RestartNode(s.dave)
+	require.NoError(ht, restartAlice(), "restart alice failed")
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_switch_test.go:206]: here the restart a bit different from `ht.RestartNode` as it's coming from `restartAlice := ht.SuspendNode(s.alice)`. So what we want to do is suspend the node first and then restart it later.

[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_switch_test.go:206]: here the restart a bit different from `ht.RestartNode` as it's coming from `restartAlice := ht.SuspendNode(s.alice)`. So what we want to do is suspend the node first and then restart it later.

File: lnwallet/channel.go [hunk 26]
```diff
@@ -3565,9 +3565,12 @@ func (lc *LightningChannel) validateCommitmentSanity(theirLogCounter,
 	// balance to satisfy the final evaluated HTLC's.
 	switch {
 	case int64(ourBalance) < 0:
-		return ErrBelowChanReserve
+		return fmt.Errorf("%w: negative local balance",
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lnwallet/channel.go:3568]: More like being easier to read from the logs,
```
commitment transaction dips peer below chan reserve: negative local balance
commitment transaction dips peer below chan reserve: our balance below chan reserve
```

Not a strong reason tho. And I'm fine either way, so let me know what you think.

File: .golangci.yml (hunk 21 above)
Comment (yyforyongyu) [reply .golangci.yml:97]: Glad you asked! This is simply from the comments below,
```
lntest/itest/lnd_switch_test.go:24:1: Duplicate words (X) found (dupword)
//  2. -------X       X         X   Bob  restart sender and intermediaries
```

And other similar places. My take is that we can skip linting those places(using `nolint:dupword`), or, we can only check for common words only. Took the lazy approach here, but I was NOT sure about it.

File: lntest/itest/lnd_mpp_test.go [hunk 27]
```diff
@@ -371,3 +293,216 @@ func (c *mppTestContext) buildRoute(ctxb context.Context, amt btcutil.Amount,
...
+	return routeResp.Route
+}
+
+// updatePolicy updates a Dave's global channel policy and returns the expected
+// policy for further check.
+func (m *mppTestScenario) updateDaveGlobalPolicy() *lnrpc.RoutingPolicy {
+	const (
+		baseFeeMsat = 500000
+		feeRate     = 0.001
+		maxHtlcMsat = 133650000
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_mpp_test.go:364]: yeah I don't understand where this is coming from either, maybe it's `calculateMaxHtlc(funding.MaxBtcFundingAmount)`?

File: lntemp/harness.go [hunk 28]
```diff
@@ -1145,36 +1144,59 @@ func (h *HarnessTest) FundCoinsP2TR(amt btcutil.Amount,
...
 	// Launch all payments simultaneously.
 	for _, payReq := range paymentRequests {
 		payReqCopy := payReq
-		wg.Add(1)
 		go send(payReqCopy)
 	}
 
 	// Wait for all payments to report success.
-	wg.Wait()
+	timer := time.After(DefaultTimeout)
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness.go:1297]: Cool I like the new helper you created! However there's more fix to this method in the following PR. To make things clear, I cherry-picked it in this PR so we can see what this method is refactored into in the last commit.

File: lnwallet/channel.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (positiveblue) [reply lnwallet/channel.go:3568]: It is good as it is, no problem

**Review by positiveblue — APPROVED** [review]
This looks pretty good @yyforyongyu thank you for taking care of it :100: 

About the linters, it will take a while to upgrade all the current code to ensure that it follows all the rules. For old code it does not matter, but if you modify it you may have to fix surrounding lines (it happened to me in another PR). It is not a bid deal though so do not worry. 

 The dupword I would leave it enable by default and use `//nolint` wherever we need, but it is mostly in comments so even if we do not catch it there is no problem deploying that code :+1:

I like a lot the new way to do itests 10/10 :fire:

**Comment by yyforyongyu** [comment]
@positiveblue cool the `dupword` is added back!

**Review by GeorgeTsagk — APPROVED** [review]
LGTM!!

very clean work! :rocket: :santa:

**Comment by yyforyongyu** [comment]
All the test failures seen in this PR have been fixed eventually when the dependant fixe PRs are merged in #6825 .

**Pull request merged**